File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    
    vector<int> topKFrequent(vector<int>& nums, int k) {
        
        int n = nums.size();
        unordered_map<int,int> count;
        for(auto ele:nums){
            count[ele]++;
        }
        
        vector<vector<int>> bucket(n+1);
        // i = freq
        // value = vector of elements having freq i
        for(auto it:count){
            int value = it.first;
            int freq = it.second;
            bucket[freq].push_back(value);
        }
        
        //iterate through the bucket from right to left
        vector<int> ans;
        for(int i = n; i > 0 ; i--){
            if(bucket[i].size()==0){
                continue;
            }
            while(bucket[i].size() > 0 && k > 0 ){
               ans.push_back(bucket[i].back());
               bucket[i].pop_back();
                k--;
            }
        }
        return ans;
    }

    
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        ans = []
        pq = []
        freq = Counter(nums)
        for key,val in freq.items():
            heapq.heappush(pq,(val,key))

            if len(pq) > k:
                heapq.heappop(pq)
        
        for count,key in pq:
            ans.append(key)
        
        return ans


        